go/database/mpt/verification.go

```diff
@@ -570,6 +612,61 @@ func verifyHashesStoredWithParents[N any](
 	return nil
 }
 
+// verifyContractCodes runs list of validation checks on accounts and on the
+// code file stored in the given directory. These checks include:
+// 1) Fatal checks
+// - All CodeHashes within accounts are present in the code file
+// - All CodeHashes within the code file are correct matching the contract byte-codes
+// 2) Non-fatal checks
+// - There are no extra Code Hashes not referenced by any account
+func verifyContractCodes(directory string, source *verificationNodeSource, observer VerificationObserver) error {
+	observer.Progress(fmt.Sprintf("Checking contract codes ..."))
+
+	codeFile := filepath.Join(directory, "codes.dat")
+	codes, err := readCodes(codeFile)
+	if err != nil {
+		return err
+	}
+
+	checkedHashes := make(map[common.Hash]bool)
+	check := func(acc *AccountNode) error {
+		codeHash := acc.info.CodeHash
+		// skip accounts that are not contracts
+		if codeHash == emptyCodeHash {
+			return nil
+		}
+		// no need to check the hash correctness more than once
+		if _, isChecked := checkedHashes[codeHash]; isChecked {
+			return nil
+		}
+		// mark an already checked hash
+		checkedHashes[codeHash] = true
+		// check that the code hash is present in the code file
+		byteCode, exists := codes[codeHash]
+		if !exists {
+			return fmt.Errorf("hash %x is missing in code file", codeHash)
+		}
+		// check correctness of the code hash
+		if got, want := common.Keccak256(byteCode), &codeHash; got.Compare(want) != 0 {
+			return fmt.Errorf("unexpected code hash for address, got: %x want: %x", got, want)
+		}
```

Comment:
I would suggest to do this check once, before walking through the accounts, for all codes in the `codes` map loaded from the file. This simplifies the code structure and you can get rid of the `checkedHashes` map.

This would also eliminate the need for checking the hashes of codes not referenced by accounts below.

Reply:
I believe it would remove check for any extra hash hence removing reporting it.

Reply:
If we were to do it before iterating accounts I think it would add another loop which is not necessary

Reply:
What is the problem with an extra loop?

It simplifies the code and eliminates extra data structures.